utils/context.helpers.js

```diff
@@ -0,0 +1,59 @@
+const lookupKeypath = (obj, keyPath) => {
+  var out = obj
+  keyPath.forEach(k => { out = out ? out[k] : undefined })
+  return out
+}
+
+// convert an array of keys to a path as used in express-validator
+const errorPath = (keys) => {
+  const [first, ...rest] = keys
+
+  const paths = rest.map(key => typeof key === 'number' ? `[${key}]` : `.${key}`)
+
+  return first + paths.join('')
+}
+
+// we define our helper functions inside the middleware because
+// we need them to access `res.local`.
+const contextMiddleware = (req, res, next) => {
+  console.log('contextMiddleware')
+
+  // a request-global stack of key paths into res.locals.
+  // these should always be called as a pair - similar to do...end.
+  // we would ideally use a caller() macro for this, but those are...
+  // unreliable at best.
+  const keyPath = []
+  res.locals.enterContext = (key) => { keyPath.push(key) }
```

Comment:
I won't have a chance to dive into this until later in the week - what's going on with enter and exit content here?

Reply:
I suggest you have a look at the tests, I tried to make it clear there :]. Basically, they maintain this key path, which affects the "context" for `getData`, `getError`, and `getName`. This way, we can push and pop to enter and exit certain branches of the data tree. This is important because now we can have nested data in the session / body, and we need a generic way to traverse it that integrates with our form controls.

Reply:
Okay will take a look when I can